### PR TITLE
fix infinite recursion in typeRel

### PR DESCRIPTION
**Pull request by jcosborn**

### CODE REVIEW

**Comment by Clyybber** [comment]
@jcosborn FYI the snippet I mentionede here https://github.com/nim-lang/Nim/pull/15211#discussion_r479564544 isn't supposed to compile, its supposed to say `(13, 23) Error: expression 'prepareTileset(parsedTileset)' has no type (or is ambiguous)`

You can specify that in the test file like so:
```nim
discard """
  errormsg: "expression 'prepareTileset(parsedTileset)' has no type (or is ambiguous)"
  line: 13
"""
```

**Comment by jcosborn** [comment]
It works fine for me (no error) when I run the test on its own.  I don't know why CI fails when it is included in megatest.

**Comment by Clyybber** [comment]
Oh, yeah sorry. You are right it should compile just fine.

**Comment by jcosborn** [comment]
`type TilesetCT[n: static[int]] = int` seems to make CI fail.  It works fine in my own tests.
Changing it to `type TilesetCT[n: static[int]] = distinct int` seems to work with CI.

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -1091,7 +1091,7 @@ proc typeRel(c: var TCandidate, f, aOrig: PType,
 
   template considerPreviousT(body: untyped) =
     var prev = PType(idTableGet(c.bindings, f))
-    if prev == nil: body
+    if prev == nil or prev == f: body
```
Comment (Araq) [inline compiler/sigmatch.nim]: Your PR is great! Unfortunately I don't understand it. How could it be that ``prev == f``?

Comment (jcosborn) [reply compiler/sigmatch.nim]: Good point.  I had only fixed the symptom.  `TilesetCT` was being compared against itself and `trDontBind` wasn't being preserved.  This is now fixed.

**Pull request merged**